bot/admin/web/src/app/faq/common/faq-definition.service.ts

```diff
@@ -0,0 +1,167 @@
+import {FaqDefinition} from './model/faq-definition';
+import {FaqSearchQuery} from './model/faq-search-query';
+import {RestService} from "../../core-nlp/rest/rest.service";
+import {map, takeUntil} from "rxjs/operators";
+import {FaqDefinitionResult} from "./model/faq-definition-result";
+import {StateService} from "../../core-nlp/state.service";
+
+/**
+ * Faq Definition operations for FAQ module
+ */
+@Injectable()
+export class FaqDefinitionService {
+
+  appIdByAppName = new Map<string, string>(); // for mock purpose only
+
+  faqData: FaqDefinitionResult = new FaqDefinitionResult([], 0, 0, 0);
+
+  constructor(private rest: RestService, private state: StateService) {
+  }
+
+  // add random data at initialization until real backend is there instead
+  setupData({applicationId, applicationName, language}:
+              { applicationId: string, applicationName: string, language: string }): void {
+
+    this.appIdByAppName.set(applicationName, applicationId);
+
+    // when there is already data for given bot / language
+    if (this.faqData.rows.some((fq: FaqDefinition) => fq.applicationId === applicationId && fq.language == language)) {
+      // no need to add mock data
+      return;
+    }
+  }
+
+  delete(fq: FaqDefinition, cancel$: Observable<any> = empty()): Observable<boolean> {
+    let newFq: FaqDefinition | undefined;
+
+    return this.rest.delete(`/faq/${fq.id}`).pipe(
+      takeUntil(cancel$),
+      map(r => {
+        //delete to current state
+        this.state.resetConfiguration()
+        if (r) {
+          this.faqData.rows = this.faqData.rows.map(item => {
+            if (fq.id && item.id === fq.id) {
+              newFq.status = 'deleted';
+              newFq = JSON.parse(JSON.stringify(fq)); // deep copy
+              return newFq;
+            } else {
+              return item;
+            }
+          });
+          return r
+        }
+      }))
+  }
+
+  save(faq: FaqDefinition, cancel$: Observable<any> = empty()): Observable<FaqDefinition> {
+    let dirty = false;
+
+    this.faqData.rows
+      .filter(item => item.id == faq.id)
+      .some(item => {
+        if (FaqDefinitionService.compareFaqSave(faq, item)) {
+          console.log(FaqDefinitionService.compareFaqSave(faq, item))
+          dirty = true
+        }
+      });
+
+    if (!dirty) {
+      return this.rest.post("/faq", faq)
+        .pipe(
+          takeUntil(cancel$),
+          map(_ => {
+            // add the current save to the state
+            this.state.resetConfiguration()
+            return JSON.parse(JSON.stringify(faq));
+          }));
+    } else {
+      return of(faq);
+    }
+  }
+
+  /**
+   * Compare faq but without creation or update Date
+   * @param newFaq
+   * @param oldFaq
+   * @private
+   */
+  private static compareFaqSave(newFaq: FaqDefinition, oldFaq: FaqDefinition): boolean {
+    return (newFaq.id == oldFaq.id
+      && newFaq.intentId == oldFaq.intentId
```

Comment:
il vaudrait mieux utiliser l operateur === 
voir : https://www.tektutorialshub.com/typescript/strict-equality-loose-equality-in-typescript/